EIPS/eip-1474.md

```diff
@@ -0,0 +1,2314 @@
+---
+eip: 1474
+title: Remote procedure call specification
+author: Paul Bouchon <mail@bitpshr.net>
+discussions-to: https://ethereum-magicians.org/t/eip-remote-procedure-call-specification/1537
+status: Draft
+type: Standards Track
+category: Interface
+created: 2018-10-02
+---
+
+## Simple Summary
+
+This proposal defines a standard set of remote procedure call methods that an Ethereum node should implement.
+
+## Abstract
+
+Nodes created by the current generation of Ethereum clients expose inconsistent and incompatible remote procedure call (RPC) methods because no formal Ethereum RPC specification exists. This proposal standardizes such a specification to provide developers with a predictable Ethereum RPC interface regardless of underlying node implementation.
+
+## Specification
+
+### Concepts
+
+#### RFC-2119
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+#### JSON-RPC
+
+Communication with Ethereum nodes is accomplished using [JSON-RPC](https://www.jsonrpc.org/specification), a stateless, lightweight [remote procedure call](https://en.wikipedia.org/wiki/Remote_procedure_call) protocol that uses [JSON](http://www.json.org/) as its data format. Ethereum RPC methods **MUST** be called using [JSON-RPC request objects](https://www.jsonrpc.org/specification#request_object) and **MUST** respond with [JSON-RPC response objects](https://www.jsonrpc.org/specification#response_object).
+
+#### Error codes
+
+If an Ethereum RPC method encounters an error, the `error` member included on the response object **MUST** be an object containing a `code` member and descriptive `message` member. The following list contains all possible error codes and associated messages:
+
+|Code|Message|Meaning|Category|
+|-|-|-|-|
+|-32700|Parse error|Invalid JSON|standard|
+|-32600|Invalid request|JSON is not a valid request object|standard|
+|-32601|Method not found|Method does not exist|standard|
+|-32602|Invalid params|Invalid method parameters|standard|
+|-32603|Internal error|Internal JSON-RPC error|standard|
+|-32000|Invalid input|Missing or invalid parameters|non-standard|
+|-32001|Resource not found|Requested resource not found|non-standard|
+|-32002|Resource unavailable|Requested resource not available|non-standard|
+|-32003|Transaction rejected|Transaction creation failed|non-standard|
+|-32004|Method not supported|Method is not implemented|non-standard|
+
```

Comment:
Parity uses a whole suite of error codes that aren't really covered here, but they don't seem to be uniformly implemented. Some references: https://github.com/paritytech/parity-ethereum/blob/80a83c95d3168c8af18e492f79657fd69558821a/rpc/src/v1/helpers/errors.rs#L37
https://github.com/ethereum/go-ethereum/issues/19027

Reply:
@MrChico 
Haven't "account is locked" error already implemented with code -32000 here -
https://ethereum.stackexchange.com/questions/8478/account-is-locked-how-to-unlock-it-using-json-rpc
But in the docs we have -32000 as "Invalid Input" error, here -
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1474.md#specification
I am confused.